### PR TITLE
Rewrite KhaveeVerifier: Fix the logic of check_sara, check_marttra, is_sumpus, handle_karun_sound_silence, and more

### PR DESCRIPTION
**Pull request by Warit-Yuv**
## Description

This pull request addresses several critical defects and bugs in the Thai poetry/rhyme processing submodule `pythainlp/khavee/core.py`. The modifications improve vowel (`check_sara`) detection, syllable spelling section (`check_marttra`) classification, rhyming logic (`is_sumpus`), and silent suffix ตัวการันต์ stripping (`handle_karun_sound_silence`) to ensure strict compliance with standard Thai grammar and poetic constraints.

### 1. Complex Silent Suffix Handling (`handle_karun_sound_silence`)

* **Problem:** The previous hardcoded string truncation logic (`word[:-2]`) assumes that all silent markers cover exactly one consonant + one Karun (-์). It fails entirely on multi-letter silent suffixes or clusters containing quiet vowels (e.g., ธุ์, ธิ์, ตริย์), leaving leftover unpronounced letters that disrupt vowel and syllable checks.
* **Solution:** Expanded the function to properly strip complex, multi-character silent clusters before evaluating vowels or spelling sections.

```python
# Fixed edge cases for silent suffixes:
kv.handle_karun_sound_silence("จันทร์")      # "จันท" -> "จัน"
kv.handle_karun_sound_silence("สิทธิ์")       # "สิทธ" -> "สิท"
kv.handle_karun_sound_silence("กษัตริย์")     # "กษัตริ" -> "กษัต"
kv.handle_karun_sound_silence("พระลักษมณ์") # "พระลักษม" -> "พระลัก"
```

### 2. Initial Cluster and Semi-Vowel Separation (`_is_true_final`)

* **Problem:** The deprecated internal method `_has_true_final_yl` did not check for tone marks (วรรณยุกต์), failing on target words like `ใกล้` because the string ends with a diacritic (วรรณยุกต์) instead of a consonant. It also misclassified instances where a letter functioned as a vowel component rather than a true final consonant ตัวสะกด (e.g., `เสีย`, `ไทย`).
* **Solution:** Added a unified internal function `_is_true_final` that strips ตัวการันต์ and วรรณยุกต์ first, then systematically filters out native Thai **คำควบกล้ำ** (true clusters) and **อักษรนำ** (leading consonants) paired with pre-posed vowels (`เ-`, `แ-`, `โ-`, `ไ-`, `ใ-`). This prevents semi-vowels and liquids (`ล`, `ร`, `ว`) functioning as initial consonant clusters from being falsely flagged as final consonants.
```python
words_to_test = ["เสีย","ไกล","ไทย","ใกล้","ไกว","ใคร","แปร"]
for word in words_to_test:
    print(f"Word: {word}, true final: {kv._has_true_final_yl(word)}, Sara: {kv.check_sara(word)}, Marttra: {kv.check_marttra(word)}")
"""
# Previous:
Word: เสีย, true final: True, Sara: เอีย, Marttra: กา
Word: ไกล, true final: True, Sara: ไอ, Marttra: เกย
Word: ไทย, true final: True, Sara: ไอ, Marttra: เกย
Word: ใกล้, true final: False, Sara: ไอ, Marttra: เกย
Word: ไกว, true final: False, Sara: ไอ, Marttra: กา
Word: ใคร, true final: False, Sara: ไอ, Marttra: กา
Word: แปร, true final: False, Sara: แอ, Marttra: กน
# After:
Word: เสีย, true final: False, Sara: เอีย, Marttra: กา
Word: ไกล, true final: False, Sara: ไอ, Marttra: กา
Word: ไทย, true final: False, Sara: ไอ, Marttra: กา
Word: ใกล้, true final: False, Sara: ไอ, Marttra: กา
Word: ไกว, true final: False, Sara: ไอ, Marttra: กา
Word: ใคร, true final: False, Sara: ไอ, Marttra: กา
Word: แปร, true final: False, Sara: แอ, Marttra: กา
"""
```

### 3. Syllable Spelling Section Corrections (`check_marttra`)

* **Problem:** Words disguised by initial consonant clusters (คำควบกล้ำ) wrapped in pre-posed vowels (e.g., `เขว`, `แปร`, `ไถล`) were previously misclassified into incorrect closed-syllable spelling sections มาตราตัวสะกด (e.g., แม่**เกอว**, แม่**กน**) instead of แม่ ก **กา**. Standalone words representing individual letters (e.g., ธ, ณ) were also assigned to closed categories.
* **Solution:** Updated `check_marttra` to verify tokens against `_is_true_final`. Lexical clusters like `โปร`, `แปร`, `ไกล`, `ไกว`, and `เปล` are now correctly routed into the open-syllable **แม่ ก กา** category. Standalone characters like `ธ`, `ณ`, `พณ`, and `บ` are also normalized to **แม่ ก กา**.
* **Note on Loanwords:** English loanwords ending in an "L" sound (`ล`) (e.g., `แอปเปิล`, `โมเดล`, `เลเวล`, `ฟุตบอล`) will continue to default to **แม่กน** under this orthographic checker. Improvement could be made to handle poetic variations that treat these as **แม่เกอว** for rhyming purposes.

```python
# Before
kv.check_marttra("พาล") -> "เกย" (Misclassified final liquid)
kv.check_marttra("ธ")   -> "กด"  (Incorrect assignment of standalone token)

# After
kv.check_marttra("พาล") -> "กน"  (Correct final mapping)
kv.check_marttra("ธ")   -> "กา"  (Correct open syllable mapping)
```

### 4. Vowel Invariant and Transformation Rules (`check_sara`)

* **Problem:** Sequential loop conditions shadowed each other (causing bugs like the `"อัว"` parsing defect), and unsafe list operations were applied when merging raw `"เ"` characters into `"แ"`. 
* **Solution:** Standardized the processing loop to cleanly handle compound vowels, fixed the shadowing logic, and added fast substring constraints for the variant realizations of `"ฤ"` and `"ฦ"` (เออ, อิ, อึ).
* **Transformed and Reduced Vowels:** Integrated complete mapping conditions for reduced vowels (*สระลดรูป*) and transformed architectures (*สระเปลี่ยนรูป*) using the *Mai Tai Khu* diacritic (`-็`). This fixes errors for common vocabularies:

```python
# Handling สระลดรูป and สระเปลี่ยนรูป ensures that the vowel sounds are correctly identified for proper rhyming and syllable classification.
words_to_test = ["เลย", "พวก", "เจ็ด", "แข็ง", "ก็"]
for word in words_to_test:
    print(f"Word: {word}, Sara: {kv.check_sara(word)}, Marttra: {kv.check_marttra(word)}")
"""
# Previous:
Word: เลย, Sara: เอ, Marttra: เกย
Word: พวก, Sara: อัว, Marttra: กก
Word: เจ็ด, Sara: เอ, Marttra: กด
Word: แข็ง, Sara: แอ, Marttra: กง
Word: ก็, Sara: ออ, Marttra: กา
# After:
Word: เลย, Sara: เออ, Marttra: เกย
Word: พวก, Sara: อัว, Marttra: กก
Word: เจ็ด, Sara: เอะ, Marttra: กด
Word: แข็ง, Sara: แอะ, Marttra: กง
Word: ก็, Sara: เอาะ, Marttra: กา
```

### 5. Control Flow Refactoring for Rhyme Matching (`is_sumpus`)

* **Problem:** The sequential `elif` structure allowed arguments to short-circuit early, which caused asymmetric matching failures if `word1` normalized but `word2` did not.
* **Solution:** Replaced the conditional tree with isolated `if` statements so both inputs undergo normalization prior to final comparison. This cleanly aligns variations like `"อะ"` + `"แม่เกย"` into `"ไอ"` + `"แม่กา"`, and bridges the poetic rhyming concordance between `"อรรม"`, `"อัม"`, and `"อำ"` (e.g., `ธรรม` - `สัม` - `จำ`).

---

## Verification and Test Cases

The updated core class execution demonstrates proper categorization across vowel mutation, syllable validation, and cluster handling:

```python
# Syllable, Cluster, & Rhyme Validation
print(kv.check_marttra("พาล"))       # Output: "กน"
print(kv.check_marttra("ไกล"))       # Output: "กา"
print(kv.check_marttra("ไกว"))       # Output: "กา"
print(kv.check_marttra("โปร"))       # Output: "กา"

print(kv.is_sumpus("บ้าน", "พาล"))  # Output: True
print(kv.is_sumpus("ธรรม", "สัม"))  # Output: True
print(kv.is_sumpus("ธรรม", "จำ"))   # Output: True
print(kv.is_sumpus("กัย", "ไก"))    # Output: True
print(kv.is_sumpus("ใจ", "ไทย"))    # Output: True

# Silent Suffix Truncation Evaluation
karun_examples = ["อินทรีย์", "กษัตริย์", "ไปรษณีย์", "ฤทธิ์", "จันทร์", "ภาพยนตร์", "กาสาวพัสตร์"]
for word in karun_examples:
    print(f"{word} -> {kv.handle_karun_sound_silence(word)}")
```

### CODE REVIEW

File: pythainlp/khavee/core.py [hunk 1]
```diff
@@ -250,20 +316,26 @@ def check_marttra(self, word: str) -> str:
...
-        # For words with ไ/ใ, check if ย/ล is a true final or just part of vowel
+        # Check for ไ/ใ
         if "ไ" in word or "ใ" in word:
-            if word[-1] not in ["ย", "ล"]:
+            if word[-1] not in ["ย", "ล", "ร", "ว"]:
                 return "กา"
-            elif not self._has_true_final_yl(word):
-                # ย/ล is part of the vowel sound, not a true final
+            elif not self._is_true_final(word):
+                return "กา"
```
[the remark this answers is not in the document]
Comment (Warit-Yuv) [reply pythainlp/khavee/core.py:521]: I have already fix the test case. Also, I have add more test case to better cover many edge case.

File: pythainlp/khavee/core.py [hunk 2]
```diff
@@ -249,28 +354,56 @@ def check_marttra(self, word: str) -> str:
...
+        if any(word.endswith(ex) for ex in silent_vowel_exceptions):
+            word = word[:-1]
+
+        # Check for อักษรตัวเดียวแทนคำ Standalone words
+        if word in ["บ", "ณ", "ธ", "พณ", "ฤ", "ฦ"]:
+            return "กา"
+
+        # Check for ำ or นิคหิต (-ํ) + า
+        if word[-1] == "ำ" or word.endswith("ํา"):
+            return "กา"
```
[the remark this answers is not in the document]
Comment (Warit-Yuv) [reply pythainlp/khavee/core.py:503]: The current implementation is already correct here.

According to the Royal Society of Thailand (ราชบัณฑิตยสภา) อำ, ไอ, ใอ, เอา are officially classified as สระเกิน. Although   phonetically they have the sound of the อะ + แม่กม, อะ + แม่เกย, and อะ + แม่เกอว, when classify Mattra มาตราตัวสะกด we only focus on the Orthography/Spelling (รูป) and not on the Phonology/Sound (เสียง). This makes word with อำ, ไอ, ใอ, เอา classified as แม่ ก "กา".
This could be further seen in บทกลอนแม่ ก กา from หนังสือ ประถม ก กา (ฉบับหอสมุดแห่งชาติ).

Thus words ending with สระ"อำ" (as well as ไอ, ใอ, เอา, ฤ, ฤๅ, ฦ, ฦๅ) is classified as แม่ ก "กา" in มาตราตัวสะกด `check_mattra`.

However, words ending with "-ำ" must rhyme with อัม (อะ + แม่กม). this logic also apply "ไอ" and "ใอ" that must rhyme with "อัย". ("เอา" has no equivalent อะ + แม่เกอว because อัว is สระอัว not เอา)  Thus, we need to normalize these word in `is_sumpus`
```# อัย -> ไอ (Normalize 'อะ' + 'เกย' to 'ไอ' + 'กา')
        if sara1 == "อะ" and marttra1 == "เกย":
            sara1 = "ไอ"
            marttra1 = "กา"
        if sara2 == "อะ" and marttra2 == "เกย":
            sara2 = "ไอ"
            marttra2 = "กา"
        # อัม -> อำ (Normalize both 'อำ' and 'อะ' + 'กม' to 'อำ' + 'กา')
        if (sara1 == "อะ" or sara1 == "อำ") and marttra1 == "กม":
            sara1 = "อำ"
            marttra1 = "กา"
        if (sara2 == "อะ" or sara2 == "อำ") and marttra2 == "กม":
            sara2 = "อำ"
            marttra2 = "กา"
```

File: tests/core/test_khavee.py [hunk 3]
```diff
@@ -71,25 +300,205 @@ def test_check_marttra(self):
...
+        self.assertEqual(kv.check_marttra("เอาะ"), "กา")
+        self.assertEqual(kv.check_marttra("เออ"), "กา")
+        self.assertEqual(kv.check_marttra("เอียะ"), "กา")
+        self.assertEqual(kv.check_marttra("เอีย"), "กา")
+        self.assertEqual(kv.check_marttra("เอือะ"), "กา")
+        self.assertEqual(kv.check_marttra("เอือ"), "กา")
+        self.assertEqual(kv.check_marttra("อัวะ"), "กา")
+        self.assertEqual(kv.check_marttra("อัว"), "กา")
+        self.assertEqual(kv.check_marttra("อำ"), "กา")
+        self.assertEqual(kv.check_marttra("ไอ"), "กา")
```
[the remark this answers is not in the document]
Comment (Warit-Yuv) [reply tests/core/test_khavee.py:413]: The current implementation is already correct here. The test case is correct in `check_mattra`.

According to the Royal Society of Thailand (ราชบัณฑิตยสภา) อำ, ไอ, ใอ, เอา are officially classified as สระเกิน. Although   phonetically they have the sound of the อะ + แม่กม, อะ + แม่เกย, and อะ + แม่เกอว, when classify Mattra มาตราตัวสะกด we only focus on the Orthography/Spelling (รูป) and not on the Phonology/Sound (เสียง). This makes word with อำ, ไอ, ใอ, เอา classified as แม่ ก "กา".
This could be further seen in บทกลอนแม่ ก กา from หนังสือ ประถม ก กา (ฉบับหอสมุดแห่งชาติ).

Thus words ending with สระ"อำ" (as well as ไอ, ใอ, เอา, ฤ, ฤๅ, ฦ, ฦๅ) is classified as แม่ ก "กา" in มาตราตัวสะกด `check_mattra`.

However, words ending with "-ำ" must rhyme with อัม (อะ + แม่กม). this logic also apply "ไอ" and "ใอ" that must rhyme with "อัย". ("เอา" has no equivalent อะ + แม่เกอว because อัว is สระอัว not เอา)  Thus, we need to normalize these word in `is_sumpus`
```# อัย -> ไอ (Normalize 'อะ' + 'เกย' to 'ไอ' + 'กา')
        if sara1 == "อะ" and marttra1 == "เกย":
            sara1 = "ไอ"
            marttra1 = "กา"
        if sara2 == "อะ" and marttra2 == "เกย":
            sara2 = "ไอ"
            marttra2 = "กา"
        # อัม -> อำ (Normalize both 'อำ' and 'อะ' + 'กม' to 'อำ' + 'กา')
        if (sara1 == "อะ" or sara1 == "อำ") and marttra1 == "กม":
            sara1 = "อำ"
            marttra1 = "กา"
        if (sara2 == "อะ" or sara2 == "อำ") and marttra2 == "กม":
            sara2 = "อำ"
            marttra2 = "กา"
```

File: tests/core/test_khavee.py [hunk 4]
```diff
@@ -71,25 +300,205 @@ def test_check_marttra(self):
...
+        self.assertEqual(kv.check_marttra("เกียร์"), "กา")
+        self.assertEqual(kv.check_marttra("เอือ"), "กา")
+        self.assertEqual(kv.check_marttra("เสือ"), "กา")
+        self.assertEqual(kv.check_marttra("เขือ"), "กา")
+        self.assertEqual(kv.check_marttra("กลัว"), "กา")
+        self.assertEqual(kv.check_marttra("ก็"), "กา")
+        self.assertEqual(kv.check_marttra("ขอ"), "กา")
+        self.assertEqual(kv.check_marttra("งอ"), "กา")
+        self.assertEqual(kv.check_marttra("ขำ"), "กา")
+        self.assertEqual(kv.check_marttra("จำ"), "กา")
```
[the remark this answers is not in the document]
Comment (Warit-Yuv) [reply tests/core/test_khavee.py:459]: The current implementation is already correct here. The test case is correct in check_mattra.

According to the Royal Society of Thailand (ราชบัณฑิตยสภา) อำ, ไอ, ใอ, เอา are officially classified as สระเกิน. Although   phonetically they have the sound of the อะ + แม่กม, อะ + แม่เกย, and อะ + แม่เกอว, when classify Mattra มาตราตัวสะกด we only focus on the Orthography/Spelling (รูป) and not on the Phonology/Sound (เสียง). This makes word with อำ, ไอ, ใอ, เอา classified as แม่ ก "กา".
This could be further seen in บทกลอนแม่ ก กา from หนังสือ ประถม ก กา (ฉบับหอสมุดแห่งชาติ).

Thus words ending with สระ"อำ" (as well as ไอ, ใอ, เอา, ฤ, ฤๅ, ฦ, ฦๅ) is classified as แม่ ก "กา" in มาตราตัวสะกด `check_mattra`.

However, words ending with "-ำ" must rhyme with อัม (อะ + แม่กม). this logic also apply "ไอ" and "ใอ" that must rhyme with "อัย". ("เอา" has no equivalent อะ + แม่เกอว because อัว is สระอัว not เอา)  Thus, we need to normalize these word in `is_sumpus`
```# อัย -> ไอ (Normalize 'อะ' + 'เกย' to 'ไอ' + 'กา')
        if sara1 == "อะ" and marttra1 == "เกย":
            sara1 = "ไอ"
            marttra1 = "กา"
        if sara2 == "อะ" and marttra2 == "เกย":
            sara2 = "ไอ"
            marttra2 = "กา"
        # อัม -> อำ (Normalize both 'อำ' and 'อะ' + 'กม' to 'อำ' + 'กา')
        if (sara1 == "อะ" or sara1 == "อำ") and marttra1 == "กม":
            sara1 = "อำ"
            marttra1 = "กา"
        if (sara2 == "อะ" or sara2 == "อำ") and marttra2 == "กม":
            sara2 = "อำ"
            marttra2 = "กา"
```

File: pythainlp/khavee/core.py [hunk 5]
```diff
@@ -26,12 +27,75 @@ def _has_true_final_yl(self, word: str) -> bool:
         :return: True if ย or ล is a true final consonant
         :rtype: bool
         """
+        return self._is_true_final(word)
+
+    def _is_true_final(self, word: str) -> bool:
```
Comment (Warit-Yuv) [inline pythainlp/khavee/core.py:129]: It's impossible to refactor and reduce the cognitive complexity down to 15. To properly detect if the word is a true final or not really require that many logic check. Only way to reduce is to split the function into many subfunction which in my opinion is worse.

File: pythainlp/khavee/core.py [hunk 6]
```diff
@@ -51,15 +115,28 @@ def check_sara(self, word: str) -> str:
...
+
+        # In case of การันย์ (word stripped of Karun characters)
+        word = self.handle_karun_sound_silence(word)
+        # Remove tonemarks for checking endings safely
+        word_req = remove_tonemark(word)
+
+        # Intercept Pali/Sanskrit words with silent terminal vowels (สระที่ไม่ออกเสียงท้ายคำ) Removing the final character -ิ or -ุ
+        silent_vowel_exceptions = ["เกียรติ", "ชาติ", "ญาติ", "มัติ", "วัติ", "บัติ", "ญัติ", "ยัติ",
+                                   "ภูมิ", "พฤติ", "พรรดิ", "วรรดิ", "พยาธิ", "โพธิ", "เกตุ", "เมรุ", "เหตุ", "ธาตุ", "วุฒิ"]
+        if any(word_req.endswith(ex) for ex in silent_vowel_exceptions):
```
[the remark this answers is not in the document]
Comment (Warit-Yuv) [reply pythainlp/khavee/core.py]: this wouldn't work. It is not as simple as list inclusion check. These word are silent when they act as suffix. Here are some example:
พระเกียรติ (พระ-เกียด), ธรรมชาติ (ทัม-มะ-ชาด), ปฏิบัติ (ปะ-ติ-บัด)
We cannot guaranteed that the tokenizer or people who use the library will break up all the word into chunks so we can't just use `word_req in silent_vowel_exceptions`.

Noted that it only silenced when act as suffix (`endswith`) not as infix/prefix -> ชาติพันธุ์ อ่านว่า "ชาด-ติ-พัน" (ไม่ใช่ ชาด-พัน)

Therefore the use of `any(word_req.endswith(ex) for ex in silent_vowel_exceptions)` here is justified.

File: pythainlp/khavee/core.py [hunk 7]
```diff
@@ -86,31 +163,29 @@ def check_sara(self, word: str) -> str:
...
             elif i in ("ไ", "ใ"):
                 sara.append("ไอ")
             elif i == "็":
-                sara.append("ออ")
+                sara.append("็")
             elif "รร" in word:
                 if self.check_marttra(word) == "กม":
                     sara.append("อำ")
                 else:
                     sara.append("อะ")
```
[the remark this answers is not in the document]
Comment (Warit-Yuv) [reply pythainlp/khavee/core.py]: "รร" should not be evaluated on a character-by-character basis inside that loop. I will push a fix.

**Review by wannaphong — APPROVED** [review]

**Review by wannaphong — APPROVED** [review]

**Comment by wannaphong** [comment]
Thank you!

**Comment by Warit-Yuv** [comment]
@wannaphong  Please update the documents https://pythainlp.org/docs/5.3/api/khavee.html and properly credited me as a contributor.  Thankyou.

**Comment by wannaphong** [comment]
> @wannaphong Please update the documents https://pythainlp.org/docs/5.3/api/khavee.html and properly credited me as a contributor. Thankyou.

It is in https://github.com/PyThaiNLP/pythainlp/releases/tag/v5.3.6 that be bug fixes only. New functions will be in 6.0.

**Comment by Warit-Yuv** [comment]
Noted. Thankyou for the update.
Will contribute more when I'm free.
